aten/src/TH/THTensor.hpp

```diff
@@ -97,32 +101,83 @@ struct THTensor
         delete this;
       }
     }
+
+    friend const int64_t* THTensor_getSizePtr(THTensor* tensor);
+    friend const int64_t* THTensor_getStridePtr(THTensor* tensor);
+    friend void THTensor_resizeDim(THTensor* tensor, int64_t ndim);
+    friend void THTensor_setSizesAndStrides(THTensor* tensor, std::vector<int64_t>&& new_size, std::vector<int64_t>&& new_stride);
+    friend void THTensor_setSizeAtDim(THTensor* tensor, int dim, int64_t new_size);
+    friend void THTensor_setStrideAtDim(THTensor* tensor, int dim, int64_t new_stride);
+    friend void THTensor_setStorageOffset(THTensor* tensor, ptrdiff_t storage_offset);
+    friend THStorage* THTensor_getStoragePtr(const THTensor* tensor);
+    friend void THTensor_stealAndSetStoragePtr(THTensor* tensor, THStorage* storage);
 };
 
 #include "generic/THTensorFastGetSet.hpp"
 #include "THGenerateAllTypes.h"
 
-inline int64_t* THTensor_getSizePtr(THTensor* tensor) {
+inline const int64_t* THTensor_getSizePtr(THTensor* tensor) {
   return tensor->sizes_.data();
 }
 
-inline int64_t* THTensor_getStridePtr(THTensor* tensor) {
+inline const int64_t* THTensor_getStridePtr(THTensor* tensor) {
   return tensor->strides_.data();
 }
 
+// NB: This is *truly* a resize; calling code (e.g., squeeze)
+// assumes that old values are preserved
+// TODO: Get rid of this function; it's a bit inefficient from
+// a calculating cached numels perspective.
 inline void THTensor_resizeDim(THTensor* tensor, int64_t ndim) {
-  // NB: This is *truly* a resize; calling code (e.g., squeeze)
-  // assumes that old values are preserved
-  tensor->sizes_.resize(ndim);
-  tensor->strides_.resize(ndim);
+  AT_ASSERT(ndim >= 0);
+  if (ndim < static_cast<int64_t>(tensor->sizes_.size())) {
+    if (tensor->numel_ != 0) {
+      for (int64_t i = static_cast<int64_t>(tensor->sizes_.size()) - 1; i >= ndim; i--) {
```

Comment:
I'd just unify these two cases; there's no guarantee the divide strategy is better than the multiply.